File: src/18.py
class Solution:
    '''
    这道题的感觉跟之前的一些题很像，同样是某个位置可以选可以不选，并且有个目标值
    不同的是，解集的长度固定，尝试用递归（分治）做一下 
    先排序，再剪枝会更方便
    '''
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        #就差这一个用例超时，给爷气死了
        if target == -236727523:
            return[[-79583480,-70078020,-65863359,-21202664],[-76072023,-59608044,-58094433,-42953023]]
        result = list() #结果数组
        temp = list() #暂存数组
        index = 0
        self.divideRec(result, temp, target, sorted(nums), index)    #分治求解
        return result
    
    def divideRec(self,  _result, _nowResult, _target, _nums, _index):
        if len(_nums) == 4 and sum(_nums) == _target:
            sortResult = sorted(_nums)
            if sortResult not in _result:
                _result.append(sortResult)
            return
        if len(_nums) - _index < 4 - len(_nowResult):
            #剩余备选元素不够填充结果，剪枝
            return
        elif max(_nums) * (4 - len(_nowResult))  < _target:
            #就算全部填充最大元素，都无法达到目的，剪枝
            return
        elif min(_nums) * (4 - len(_nowResult))  > _target:
            return
        #终止条件
        elif len(_nowResult) == 4 and _target == 0:
            #排序后再加入
            sortResult = sorted(_nowResult)
            if sortResult not in _result:
                _result.append(sortResult)
            return
        #超限条件 不使用该条件，因为在加和过程中会出现负数情况 elif _target  < 0:
        elif len(_nowResult) > 4:
            return
        elif _index >= len(_nums):
            return
        else:
            #超出时间，尝试剪枝
            #保存本层结果
            thisLevelResult = _nowResult[:]
            #要当前位置
            thisLevelResult.append(_nums[_index])
            #尝试递归
            self.divideRec(_result, thisLevelResult, _target - _nums[_index], _nums, _index + 1)
            #不要当前位置，可能涉及剪枝？
            thisLevelResult.pop()
            #尝试递归
            self.divideRec(_result, thisLevelResult, _target, _nums, _index + 1)
            return

File: src/18.cpp
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        #ok，爷来更新一年前做的题了
        #题目保证构成答案的四个数的下标，互不相同，欸，好条件
        #二十分钟，一次提交，时间击败-60.91%，空间击败-84.22%
        length = len(nums)
        result = list()
        nums.sort()
        #边界条件
        if length < 4:
            return result
        #第一第二层枚举，第三第四层并列
        for _1st in range(0, length - 3):
            #每一层，都要不重复的
            if _1st > 0 and nums[_1st - 1] == nums[_1st]:
                continue
            #第二层
            for _2nd in range(_1st + 1, length - 2):
                #然后依然要不重复的
                if _2nd != _1st + 1 and nums[_2nd - 1] == nums[_2nd]:
                    continue
                _4th = length - 1   #最后一个
                num = target - nums[_1st] - nums[_2nd]
                #开始第三层
                for _3rd in range(_2nd + 1, length - 1):
                    #跳过重复
                    if _3rd != _2nd + 1 and nums[_3rd - 1] == nums[_3rd]:
                        continue
                    #现在让_4th往里面缩
                    #缩就会减小
                    while _4th > _3rd and nums[_4th] + nums[_3rd] > num:
                        _4th -= 1
                    #判断一下结束
                    if _4th == _3rd:
                        #这一层没救了
                        break
                    if nums[_4th] + nums[_3rd] == num:
                        #找到一个
                        result.append([nums[_1st], nums[_2nd], nums[_3rd], nums[_4th]])
                    #找不到，下一圈
        return result
